rasa/core/test.py

```diff
@@ -481,13 +489,38 @@ def emulate_loop_rejection(partial_tracker: DialogueStateTracker) -> None:
     partial_tracker.update(ActionExecutionRejected(rejected_action_name))
 
 
+def _get_processor_entity_evaluation_result(
+    processor: "MessageProcessor",
+    tracker: DialogueStateTracker,
+    prediction: PolicyPrediction,
+):
+    previous_event = tracker.events[-1]
+    if isinstance(previous_event, (UserUttered)):
+        entity_predictions = [
+            entity
+            for prediction_event in prediction.events
+            if isinstance(prediction_event, EntitiesAdded)
+            for entity in prediction_event.entities
+        ]
+        entity_targets = previous_event.entities
```

Comment:
how will it work in case of e2e testing? because I believe it'll contain both nlu entities and e2e entities

Reply:
I thought that these entities were the ones specified in the story step, e.g. `my name is [joe](name)`. What do you mean by nlu entities in this context?

Reply:
> I thought that these entities were the ones specified in the story step
yes, I'm calling `entities specified in the story step` as `e2e entities`. But there are also other entities that are extracted by nlu pipeline. I think we should test them separately

Reply:
so if the nlu pipeline extracts an entity it will also end up in `previous_event.entities`?

Reply:
it'll end up in `UserUttered` but not in `EntitiesAdded`. I think @wochinge knows this better

Reply:
Vova is right. [This](https://github.com/RasaHQ/rasa/blob/954e4734d479924df7211cd3bc5503285d54a66b/rasa/shared/core/events.py#L758) and [this](https://github.com/RasaHQ/rasa/blob/954e4734d479924df7211cd3bc5503285d54a66b/rasa/shared/core/events.py#L418) should hopefully explain things.

Reply:
but I don't think this is the case when running `test`. If you look at `_collect_user_uttered_predictions` the predicted entities are not added to the event.

Reply:
I think that's because `_clean_entity_results` is creating a new list before the action could add the `EntitiesAdded` event. I guess we have to predict the next action _before_ collecting `_collect_user_uttered_predictions` as we're are otherwise missing a piece of information.

Reply:
I don't think `_clean_entity_results` matters here. I'm not using the `user_uttered_eval_store` I'm just using what's in the `partial_tracker` which doesn't include the predicted entities, only the ones from the story.